### PR TITLE
Include corelib in rust runner binary

### LINKED ISSUE #2066: Include corelib in rust test runner binary
Currently, rust test runner corelib detection depends on either passing `--corelib-path` or hardcoded paths relative to protostar directory root. First solution is problematic and the second one will not work in case we try to run test runner in any other context than its current directory in protostar repository.

Until #1997 is resolved, we have to figure out a solution for bundling `corelib` with the runner binary so that it works without passing any arguments.

New CI tasks should be introduced as well for building the binary.

### PR DESCRIPTION
**Pull request by cptartur**
Closes #2066 

### Checklist
- [ ] Relevant issue is linked
- [ ] Docs updated/issue for docs created
- [ ] Added relevant tests

### CODE REVIEW

**Review by piotmag769 — APPROVED** [review]
LGTM

**Review by karol-bisztyga — CHANGES_REQUESTED** [review]
blocked by https://github.com/software-mansion/protostar/pull/2059

**Review by piotmag769 — APPROVED** [review]

File: protostar-rust/src/main.rs [hunk 1]
```diff
@@ -1,30 +1,37 @@
...
+    let tmp_dir = tempdir()?;
+    CORELIB
+        .extract(&tmp_dir)
+        .expect("Failed to copy corelib to temporary directory");
+    Ok(tmp_dir)
 }
 
 fn main_execution() -> Result<()> {
-    let args = Args::parse();
+    let _args = Args::parse();
```
Comment (karol-bisztyga) [inline protostar-rust/src/main.rs:28]: why is this not removed?

File: protostar-rust/src/main.rs [hunk 2]
```diff
@@ -1,30 +1,37 @@
...
 use clap::Parser;
+use include_dir::{include_dir, Dir};
 use scarb_metadata::MetadataCommand;
+use std::path::PathBuf;
+use tempfile::{tempdir, TempDir};
 
 use rust_test_runner::pretty_printing;
 use rust_test_runner::run_test_runner;
 
+static CORELIB: Dir = include_dir!("../cairo/corelib/src");
```
Comment (karol-bisztyga) [inline protostar-rust/src/main.rs:12]:
```suggestion
static CORELIB_PATH: Dir = include_dir!("../cairo/corelib/src");
```
unclear naming

File: protostar-rust/src/main.rs [hunk 3]
```diff
@@ -1,30 +1,37 @@
...
+    let _args = Args::parse();
 
     // TODO #1997
-    let corelib = match args.corelib_path {
-        Some(corelib) => Utf8PathBuf::from(corelib),
-        None => Utf8PathBuf::from("../../cairo/corelib/src")
-            .canonicalize_utf8()
-            .context("Failed to resolve corelib path")?,
-    };
+    let corelib_dir = load_corelib()?;
```
Comment (karol-bisztyga) [inline protostar-rust/src/main.rs:31]: Why is `load_corelib` a separate function? This code isn't reused and is not even big. I think it should be inlined here.

File: protostar-rust/src/main.rs (hunk 2 above)
Comment (karol-bisztyga) [inline protostar-rust/src/main.rs:12]: why is this static? It's used only in one place.

**Review by karol-bisztyga — CHANGES_REQUESTED** [review]

File: protostar-rust/src/main.rs (hunk 3 above)
Comment (MaksymilianDemitraszek) [reply protostar-rust/src/main.rs:31]: I am not sure tbh, it abstracts the operation nicely

Comment (karol-bisztyga) [reply protostar-rust/src/main.rs:31]: It can be brought down to literally one instruction:
```
CORELIB
        .extract(&tempdir()?)
        .expect("Failed to copy corelib to temporary directory");
```

Comment (MaksymilianDemitraszek) [reply protostar-rust/src/main.rs:31]: Yea but looking at it i am not sure what it does at first glance and function names it

Comment (karol-bisztyga) [reply protostar-rust/src/main.rs:31]: LoL, the error message tells you exactly what it does:

> Failed to copy corelib to temporary directory

which means that what it does is: "copy corelib to temporary directory". It's even more descriptive than the function name "load_corelib" (btw isn't this name misleading? "load corelib" does not indicate that we copy it to a tmp dir, it might also mean that we load this into the memory or something which brings confusion).

> I would stick with a function, this way we reduce cognitive complexity imo

Extracting single instructions into separate once-used functions looks like an anti-pattern to me. This leads to a code I'd not wish to maintain. Also, as I mentioned above, the name of this function is not quite compatible with the error message and with what's actually happening and it is misleading.

Long story short, I know it's 2v1 guys and usually in such a case I back off right away but here I find your arguments not super legit, sorry 🤷 At the end of the day I'm only an advisor here, the author may do as he wishes.

Comment (cptartur) [reply protostar-rust/src/main.rs:31]: I'd rather have multiple small functions in the codebase that make it readable. From the point of person doing something in context of `fn main`, how exactly is corelib loaded is not important at all.

File: protostar-rust/src/main.rs (hunk 2 above)
Comment (cptartur) [reply protostar-rust/src/main.rs:12]: The directory is being included in the binary at the compile time so there's no reason to do the same with the path to it.

File: protostar-rust/src/main.rs (hunk 1 above)
Comment (cptartur) [reply protostar-rust/src/main.rs:28]: Because it is being used in other PR (#1999). The CLI main interface prototype was designed for a specific purpose of handling arguments of the CLI.

**Review by karol-bisztyga — APPROVED** [review]

File: protostar-rust/src/main.rs (hunk 2 above)
Comment (karol-bisztyga) [reply protostar-rust/src/main.rs]: Discussed offline

> ... there's no reason to do the same with the path to it.

It should be ("not" missing, therefore confusion)

```
... there's no reason not to do the same with the path to it.
```
